### ISSUE TITLE
Support custom file names, including date, when saving / detaching attachments

### ISSUE THREAD

**Issue by reagle** [RealRaven2000/FiltaQuilla#219]
It'd be great if we had a format string with which we could specify the names.

**Comment by RealRaven2000** [comment]
what formats should it support?

**Comment by RealRaven2000** [comment]
> It'd be great if we had a format string with which we could specify the names.

I just realized that this is obviously a FiltaQuilla feature - so I have transferred the issue to the correct Add-on.

**Comment by reagle** [comment]
[Attachment extractor](https://addons.thunderbird.net/en-US/thunderbird/addon/attachment-extractor/?src=search) supports: `%date%, %time%, %fromMail%, %subject%, %filename%` . I like that it can rename, but I don't know how to invoke it during a filter. FiltaQilla can obviously run a detach, but doesn't support user specified names.

**Comment by RealRaven2000** [comment]
> [Attachment extractor](https://addons.thunderbird.net/en-US/thunderbird/addon/attachment-extractor/?src=search) supports: `%date%, %time%, %fromMail%, %subject%, %filename%` . I like that it can rename, but I don't know how to invoke it during a filter. FiltaQilla can obviously run a detach, but doesn't support user specified names.

I like that it uses the same syntax that SmartTemplates invented / uses. I wonder how it formats `%date%` as slashes are not valid characters for filenames. How is it invoked at the moment? What's the workflow they use?

**Comment by reagle** [comment]
I don't know, it's just another plugin I came across when looking for this functionality. I wouldn't want slashes myself, I'd want YYYYMMDD so things sort nicely. I [currently rename myself](https://gist.github.com/reagle/5a5498b3530165a2ee63a9613f766a56) with a python script, but that breaks the links to the attachments from within Thunderbird. 

I'm hoping that if FQ supported file naming, the links from messages could be preserved.

(And I detach attachments because they tend to bloat mail folders give all the PDFs and docx files.)

**Comment by RealRaven2000** [comment]
In my Add-on [SmartTemplates](https://addons.thunderbird.net/en-US/thunderbird/addon/smarttemplate4/), I have a command `%dateformat()%`  which accepts letters. Those can also used as single tokens such `%Y%` (Year) `%m%` (2digit month with leading zero) `%d%` (day of month with leading zero)   ` %H%`  (hour, 24hrs with leading zero) `%M%` (minute) etc.

![image](https://user-images.githubusercontent.com/6314674/227770553-8d2c194d-f4fb-4180-8788-8e818c6b286d.png)

So you could do something like `%dateformat(YHM)%`

maybe we can copy the logic into FiltaQuilla. The hardest part may be documenting everything in the UI (as it needs to be localizd).

**Comment by RealRaven2000** [comment]
Before I start working on this feature, we need to convert the complete settings dialogs to HTML so that they can be displayed as a content tab. Thunderbird slowly but steadily deprecates the use of xml / xhtml in UI so it is prudent to do so for Tb115 to avoid duplicated work when it becomes inaccessible to legacy Add-ons.

**Comment by ThiloteE** [comment]
Just as a side note: https://en.wikipedia.org/wiki/ISO_8601
![image](https://github.com/RealRaven2000/FiltaQuilla/assets/73715071/5f6c5e59-6546-4465-a903-ddad81203f50)
(Source: https://xkcd.com/1179)

I would therefore suggest YYYY-MM-DD

**Comment by RealRaven2000** [comment]
> Just as a side note: 
> 
> I would therefore suggest YYYY-MM-DD

of course, that's the only one that makes sense - especially if you want to sort the titles by date.

**Comment by JF1313** [comment]
I agree that this would be a terrific feature to have. It'd be helpful to have control over the formatting of adding a prefix like the date at the start of a file name such as YYYY MM DD

**Comment by RealRaven2000** [comment]
Also related #197 which has a lot of information in it - which one should be closed as duplicate so we can concentrate efforts on completing one of them?

**Comment by RealRaven2000** [comment]
Things to consider:

- filter actions only support a single action. since we are currently using the open folder button and the folder path, we somehow need to  combine both in a way that preserves both settings (location and file name) in the same string, in a convenient manner. If the parameter would look something like this...
`path,nameFormat`
.. then I could imagine 2 buttons one for folder picker and one for name format which would allow changing both parts of the string independently.
- for name parts we should use the same syntax as run program to remain consistent, which so far includes:
```
 @subject@
 @author@ 
 @recipient@
 @cclist@
 @messageid@
 @date@
```
- for date formatting I would prefer a specialized function `@dateformat()@` with formatString consisting of special letters which I have already [documented in SmartTemplates](https://smarttemplates.quickfolders.org/variables.html#customDates). Of course these would be used without the % characters which are part of the standalone variable syntax of SmartTemplates.
```
 @dateformat(formatString)@
// for example
 @dateformat(d.m.Y H:M)@
```

**Comment by salmonslaw** [comment]
I agree it would be really great to be able to set easily a pattern for file names, and very happy that you are concerned about making sure we can have both options: set the location and the file names, both are very important for me. Personally the names should look like this @date(YYYY-MM-DD_H-M)@_@author@_to_@recipient@_@subject@. And possibility to set maximum length would be great too.

**Comment by malejdj** [comment]
Dear all,

I mentioned same thing in #323 and please, is it possible to implement something for better handling with same name files?

> In many cases, I receive multiple emails with attachments that share the same filename (e.g., invoice.pdf). Would it be possible to implement a feature that renames these attachments automatically when saving?

**Comment by RealRaven2000** [comment]
> Dear all,
> 
> I mentioned same thing in [#323](https://github.com/RealRaven2000/FiltaQuilla/issues/323) and please, is it possible to implement something for better handling with same name files?

Yes I think if we encounter a same name file some sort of hash test to check if it is a true duplicate would be great if it's possible. I will find out how it could be done. bear in mind that any arbitrary automatic file access is currently not supported by the web extension API, so all of this is based on the "experimental" part (privileged code) of the Add-on. But so it any kind of filter management (custom actions / search terms) as there is currently no filters API/

**Comment by deepu2** [comment]
Hi, I was also looking to rename the file names when using "Save Message as File/Save Attachments To" include variables, something like "[yy][mm][dd] [first name] [subject].eml". So would this issue also support "Save Message as File" action?

<img width="902" height="632" alt="Image" src="https://github.com/user-attachments/assets/d68e2f3c-b11e-426d-849f-6a9d1fffab85" />

**Comment by support337** [comment]
second this! both save message as file & save attachments functions need an option to specify destination filename template.

Would be great to allow the same <code>@MESSAGEID@</code> or other existing options for destination filename as in a Run Program action.  Or maybe ability to extract some message header value for a specified key.

<code>// @MESSAGEID@ message Id </code>

this is a must have for any further automated processing for both attachments and message, i.e. the use case is save message as <code>messageId.eml</code> & save attachments as <code>messageId-file01.jpg</code> <code>messageId-file02.pdf</code> etc. 

This will allow to have a attachment-message relation already in a filename, for example for future import into a search index or document db

**Comment by RealRaven2000** [comment]
I am just in the process of converting settings to html, then we can expand these options. 

Things to think of:

filter actions only allow 1 single string as modifiable parameter, so whatever naming conventions we come up with, it all needs to be crammed into the line where the folder path is, so really it would need a clever UI to "build" the name as part of the string. Also it would need the filename itself as an optional part. it could have %filename% as a placeholder for the original file but automatically take the extension (e.g. pdf) and remove it + add to the end. filename would probably need a "maximal length" parameter:

`E:\some\path\%date%_%filename(40)%_%messageid%`

would have these components:

```js
%date% = "2025-10-08"
%filename% = "My Report from 01.10.2025.pdf"
%messageid% = "<RealRaven2000/FiltaQuilla/issues/219/3379609273@github.com>"
```

special characters such as `</\+;,` will be replaced by FiltaQuilla with _ - we could still end up with a file name that is too long and has to be truncated. The resulting file in the example would be stored at

```
E:\some\path\2025-10-08_My Report from 01.10.2025__RealRaven2000_FiltaQuilla_issues_219_3379609273@github.com_.pdf
```

For now, the file name will just be cut off at the end if it exceeds the maximum (keeping the extension name intact).

in addition to a folder picker, we would probably need a picker for appending / building these fields:
<img width="697" height="91" alt="Image" src="https://github.com/user-attachments/assets/ff92a3fd-4c3f-4a14-a14c-ea221b062fb1" />

**Comment by smartcooky99** [comment]
I have been reading through this thread trying to understand what it all means. I'm an oldie, so a lot of the posts in here might as well be in a foreign language. 
We have literally hundreds of attachments daily coming in from dozens of senders. Going though all of them to sort each attachment or group of attachments for actioning is a daily nightmare. There used to be a great Add On called Automatic Attachment Extractor that used to do exactly this but it no longer functions. 

Perhaps someone could explain, briefly, whether or not this filter/application/extension (whatever) you are working on will do what I want to be able to do.

1. Automatically detect any and all attachments in any incoming emails
2. Rename the attachments with the name of the sender (being all before the "@" in the email address... for example, an attachment called "IMG1234.jpg" from "davidsmith@email.zz" would be saved as the name "davidsmith-IMG1234.jpg"
3. Save these renamed attachments to a predetermined folder.

Thank you

**Comment by RealRaven2000** [comment]
> * Automatically detect any and all attachments in any incoming emails

it already does this, so yes

> * Save these renamed attachments to a predetermined folder.

it already does this, so yes - the folder must exist already and has to be selected when you create the filter

> * Rename the attachments with the name of the sender (being all before the "@" in the email address... for example, an attachment called "IMG1234.jpg" from "[davidsmith@email.zz](mailto:davidsmith@email.zz)" would be saved as the name "davidsmith-IMG1234.jpg"

that's one scenario this issue would cover. But it's just a special case, so what we need is a set of rules not just one hardcoded filter that simply prepends the username of the email. That's why we will have to use placeholders that extract the data from the existing email, including the subject line, the date etc. We won't just make a new method that covers your use case only.

**Comment by smartcooky99** [comment]
Thank you for the reply
So the Placeholders used by Attachment Extractor were

`%date%` ...Date in the format YYYY-MM-DD
`%fromMail%` ...Whole email address 
`%subject%` ...Entire text of subject line
`%filename%` ...Entire filename including extension

I also suggest

`%sendername%` ... the text of the email address on the left of, but not including the "@"

The main problems with Attachment Extractor were
1. It had to be run manually each time
2. You could delete unwanted placeholders, but it would not remember those settings the next time you ran it.

Of course, the bigger problem now is that it no longer works (as of sometime after TB115) 

Could those placeholders be used as follows?

In the _"Enable all items that you want to have available in the 'filter rules' dialog. Restart after changes."_ section of FiltaQuilla Preferences, a checkbox called _"Rename attachments using Smart Template"_

In the _"Perform these actions"_ section of the Filter rules, an option in the drop-down box would read _"Rename attachments to"_ and when selected, you would insert the placeholders in input box on the right

**Comment by RealRaven2000** [comment]
First I had to fix detaching attachments, see latest  version at #369. Once that is safely released on ATN we can look at the naming functions.
